parsers/ES.py

```diff
@@ -163,42 +253,29 @@ def fetch_production(
 ) -> List[dict]:
     check_valid_parameters(zone_key, session, target_datetime)
     ses = session or Session()
-    island_data = fetch_island_data(zone_key, ses, target_datetime)
+    data_mapping = DATA_MAPPING.copy()
+
+    ## Production mapping override for Canary Islands
+    # NOTE the LNG terminals are not built yet, so power generated by "gas" or "cc" in ES-CN domain is actually using oil.
+    # Recheck this every 6 months and move to gas key if there has been a change.
+    # Last checked: 2022-06-27
+    if zone_key.split("-")[1] == "CN":
+        data_mapping["gas"] = "production.oil"
+        data_mapping["cc"] = "production.oil"
+
+    if zone_key == "ES-IB-ME" or zone_key == "ES-IB-FO":
+        data_mapping["gas"] = "production.oil"
+    if zone_key == "ES-IB-IZ":
+        data_mapping["die"] = "prosuction.gas"
+
+    island_data = fetch_island_data(zone_key, ses, target_datetime, data_mapping)
     productionEventList = ProductionBreakdownList(logger)
-
     for event in island_data:
```

Comment:
It would be better if production logic was handled here, and exhange in exchange, etc.

Reply:
That parser is under development and should probably not be used as an example until it is done. It will also likely be refactored now that we are adding storage to the parser. A better example would be ENTSOE or EIA as they are fully feature complete and handle multiple zones.

Adding so much reliance on string values will make it harder to maintain and it will be easier to introduce errors.

Reply:
Ok, I will not base my code on that parser. However, I don't see how could we make this code work and raise a warning when new unknown keys are passed. I understood from a previous message of yours that this was necessary, but after re-reading it I have noticed I may have misunderstood. I think that it is a nice addition though. Could you guide me in how would you like the parser to be structured like? Should it pass a dict with the original keys (e.g. dem, die, nuc, etc...) to all the fetch_x functions, and the conversion to Electricity Maps data be done there? It would lose the ability to check that all keys have been used.

Reply:
I now see that this current implementation is a bit fragile. I can implement this other code, if you don't have a better idea.

Reply:
If we split them into 3 lists, production, exchange, and ignored it would be easy. Either we check everything in a preprocess and make sure the key are in one of the lists. Or we check it in each of the fetch function. Both have their pros and conns, doing everything in a preprocess step will make the code slightly more complex but less code duplication.

In the second approach we do it in each parser and duplicate (some of) the code but it makes it less complex and easier to understand in my opinion.

For the preprocess approach I would do something like this:

```python
  for key of keys:
    if key in production:
      # handle production
    elif key in exchange:
      # handle exchanges
    elif key == "dem"
      # handle demand
    else:
      logger.warning("Log warning about key"
```

Here is how I would do it for production with the second approach.

```python
  for key of keys:
    if key in production_map:
      # Handle production here
    elif key not in exchange_map and key not in ignored_values: # We don't need to check for dem explicitly here as it would be in ignored values, in the preprocess step we have to explicitly check it instead.
      logger.warning("Log warning about key")
```

Both avoid the string prefix by putting them in different lists though.